File: Java-CodeServer/src/main/resources/js/prompto/value/SetValue.js
var Value = require("./Value").Value;
var Integer = require("./Integer").Integer;
var SetType = require("../type/SetType").SetType;

exports.resolve = function() {
    SetType = require("../type/SetType").SetType;
};

function SetValue(itemType, items) {
    Value.call(this, new SetType(itemType));
    this.itemType = null;
    this.items = items || {};
    return this;
}

SetValue.prototype = Object.create(Value.prototype);
SetValue.prototype.constructor = SetValue;

SetValue.prototype.addAll = function(items) {
    for(var p in items) {
        this.add(items[p]);
    }
};

SetValue.prototype.toString = function() {
    var names = Object.getOwnPropertyNames(this.items);
    return "<" + names.join(", ") + ">";
};

SetValue.prototype.add = function(item) {
    this.items[item.toString()] = item;
};

SetValue.prototype.size = function() {
    var n = 0;
    for(var p in this.items) {
        n += 1;
    }
    return n;
};

SetValue.prototype.isEmpty = function() {
    for(var p in this.items) {
        return false;
    }
    return true;
};

SetValue.prototype.hasItem = function(context, item) {
    return item.toString() in this.items;
};


SetValue.prototype.getItemInContext = function(context, index) {
    if (index instanceof Integer) {
        try {
            var idx = index.IntegerValue();
            for(var p in this.items) {
                if(--idx==0)
                    return this.items[p];
            }
            throw new IndexOutOfRangeError();
        } catch (e) {
            if(e instanceof PrestoError) {
                throw e;
            } else {
                throw new InternalError(e.toString());
            }
        }
    } else
        throw new SyntaxError("No such item:" + index.toString());
};

SetValue.prototype.Add = function(context, value) {
    if (value instanceof SetValue) {
        var result = new SetValue(this.type.itemType, this.items);
        result.addAll(value.items);
        return result;
    } else {
        return Value.prototype.Add.apply(this, context, value);
    }
};

SetValue.prototype.getIterator = function(context) {
    return new SetIterator(this.items, context);
};


SetValue.prototype.equals = function(obj) {
    if(obj instanceof SetValue) {
        for(var p in this.items) {
            var v1 = this.items[p];
            var v2 = obj.items[p];
            if(v1==v2) {
                continue;
            } else if(v1==null || v2==null) {
                return false;
            } else {
                if(v1.equals) {
                    if(!v1.equals(v2)) {
                        return false;
                    }
                } else if(v2.equals) {
                    if(!v2.equals(v1)) {
                        return false;
                    }
                } else {
                    return false;
                }
            }
        }
        return true;
    } else {
        return false;
    }
};


function SetIterator(items, context) {
    this.items = items;
    this.names = Object.getOwnPropertyNames(items);
    this.context = context;
    this.index = -1;
    return this;
}

SetIterator.prototype.hasNext = function () {
    return this.index < this.names.length - 1;
}

SetIterator.prototype.next = function() {
    return this.items[this.names[++this.index]];
};

exports.SetValue = SetValue;

File: Java-Core/src/main/java/prompto/value/SetValue.java
package prompto.value;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.Set;

import prompto.compiler.CompilerUtils;
import prompto.compiler.MethodConstant;
import prompto.compiler.MethodInfo;
import prompto.compiler.Opcode;
import prompto.compiler.IOperand;
import prompto.compiler.ResultInfo;
import prompto.error.IndexOutOfRangeError;
import prompto.error.InvalidDataError;
import prompto.error.PromptoError;
import prompto.error.SyntaxError;
import prompto.expression.IExpression;
import prompto.grammar.Identifier;
import prompto.intrinsic.PromptoSet;
import prompto.runtime.Context;
import prompto.type.IType;
import prompto.type.SetType;

public class SetValue extends BaseValue implements IContainer<IValue>, IListable<SetValue> {

	Set<IValue> items = null;
	
	public SetValue(IType itemType) {
		super(new SetType(itemType));
		this.items = newSet();
	}
	
	public SetValue(IType itemType, Set<IValue> items) {
		super(new SetType(itemType));
		this.items = items;
	}

	protected Set<IValue> newSet() {
		return new PromptoSet<IValue>();
	}

	@Override
	public boolean isEmpty() {
		return items.isEmpty();
	}

	@Override
	public long length() {
		return items.size();
	}

	public Collection<IValue> getItems() {
		return items;
	}
	
	@Override
	public boolean hasItem(Context context, IValue value) throws PromptoError {
		return items.contains(value);
	}

	@Override
	public IValue getItem(Context context, IValue index) throws PromptoError {
		if (index instanceof Integer) {
			int idx = (int)((Integer)index).longValue() - 1;
			return getNthItem(idx);
		} else
			throw new SyntaxError("No such item:" + index.toString());
	}

	private IValue getNthItem(int idx) throws PromptoError {
		Iterator<IValue> it = items.iterator();
		while(it.hasNext()) {
			IValue item = it.next();
			if(idx--==0)
				return item;
		}
		throw new IndexOutOfRangeError();
	}

	@Override
	public Iterable<IValue> getIterable(Context context) {
		return items;
	}

	@Override
	public boolean equals(Object obj) {
		if(!(obj instanceof SetValue))
			return false;
		return items.equals(((SetValue)obj).items);
	}

	@Override
	public SetValue newInstance(List<IValue> values) {
		IType itemType = ((SetType)type).getItemType();
		SetValue result = new SetValue(itemType);
		result.items.addAll(values);
		return result;
	}
	
	@Override
	public String toString() {
		return items.toString();
	}

	@Override
	public IValue getMember(Context context, Identifier id, boolean autoCreate) throws PromptoError {
		String name = id.toString();
		if ("length".equals(name))
			return new Integer(items.size());
		else
			throw new InvalidDataError("No such member:" + name);
	}

	@Override
	public IValue plus(Context context, IValue value) throws PromptoError {
        if (value instanceof BaseList<?,?>)
            return this.merge(((BaseList<?,?>)value).getItems());
        else if (value instanceof SetValue)
            return this.merge(((SetValue)value).getItems());
        else
            throw new SyntaxError("Illegal: " +this.type.getId() + " + " + value.getClass().getSimpleName());
    }

	public static ResultInfo compileAdd(Context context, MethodInfo method, ResultInfo left, IExpression exp, boolean toNative) throws SyntaxError {
		// TODO: return left if right is empty (or right if left is empty and is a set)
		// create result
		ResultInfo info = CompilerUtils.newInstance(method, PromptoSet.class); 
		// add left, current stack is: left, result, we need: result, result, left
		method.addInstruction(Opcode.DUP_X1); // stack is: result, left, result
		method.addInstruction(Opcode.SWAP); // stack is: result, result, left
		IOperand oper = new MethodConstant(PromptoSet.class, "addAll", 
				Collection.class, boolean.class);
		method.addInstruction(Opcode.INVOKEVIRTUAL, oper);
		method.addInstruction(Opcode.POP); // consume returned boolean
		// add right, current stack is: result, we need: result, result, right
		method.addInstruction(Opcode.DUP); // stack is: result, result 
		exp.compile(context, method, false); // stack is: result, result, right
		oper = new MethodConstant(PromptoSet.class, "addAll", 
				Collection.class, boolean.class);
		method.addInstruction(Opcode.INVOKEVIRTUAL, oper);
		method.addInstruction(Opcode.POP); // consume returned boolean
		return info;
	}
	
	public SetValue merge(Collection<IValue> items) {
		List<IValue> result = new ArrayList<IValue>();
		result.addAll(this.items);
		result.addAll(items);
		return newInstance(result);
	}

}
